src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs

```diff
@@ -245,27 +245,67 @@ protected override void Dispose(bool disposing)
             {
                 return;
             }
-            _disposed = true;
+            try
+            {
+                if (!_disposed && disposing)
+                {
+                    CheckAsyncTaskInProgress();
+                }
+            }
+            finally
+            {
+                CloseStreamFromDispose(disposing);
+            }
+        }
 
-            // Dispose of our resources if this StreamReader is closable.
-            if (_closable)
+        private void CloseStreamFromDispose(bool disposing)
+        {
+            try
             {
-                try
+                // Dispose of our resources if this StreamReader is closable.
+                if (_closable && !_disposed)
                 {
-                    // Note that Stream.Close() can potentially throw here. So we need to
-                    // ensure cleaning up internal resources, inside the finally block.
+                    // Attempt to close the stream
+                    // Note that Stream.Close() can potentially throw here.
+                    // In this case, we still need to ensure
+                    // cleaning up internal resources, hence the finally block.
                     if (disposing)
                     {
                         _stream.Close();
                     }
                 }
-                finally
+            }
+            finally
+            {
+                _disposed = true;
+                _charPos = 0;
+                _charLen = 0;
+                base.Dispose(disposing);
+            }
+        }
+
+        public override ValueTask DisposeAsync() =>
+            GetType() != typeof(StreamReader) ?
+                base.DisposeAsync() :
+                DisposeAsyncCore();
+
+        private ValueTask DisposeAsyncCore()
+        {
+            // Same logic as in Dispose().
+            Debug.Assert(GetType() == typeof(StreamReader));
+            try
+            {
+                if (!_disposed)
                 {
-                    _charPos = 0;
-                    _charLen = 0;
-                    base.Dispose(disposing);
+                    CheckAsyncTaskInProgress();
                 }
             }
+            finally
+            {
+                CloseStreamFromDispose(disposing: true);
+            }
```

Comment:
CheckAsyncTaskInProgress isn't called in StreamWriter.DisposeAsync, it probably shouldn't be called here either. In which case, this could be simply:
```cs
            CloseStreamFromDispose(disposing: true);
```

Reply:
Actually, @Neme12 is right on https://github.com/dotnet/runtime/pull/89477#discussion_r1274526703, you should use DisposeAsync.

Reply:
will do and change streamwriter too

Reply:
CheckAsyncTaskInProgress is called in StreamWriter.DisposeAsync, it happens as part of Flush